File: frontend/lib/objects-of-issue/overview.ts
/**
 * Overview-tab summary for Objects of the Issue (Increment O1).
 *
 * Combines live progress, computed model figures and the Objects Assessment into one shape the
 * Overview tab renders directly. Nothing here is persisted.
 */

import { assessObjectsOfIssue } from '@/lib/objects-of-issue/assessment';
import { computeObjectsOfIssueModel } from '@/lib/objects-of-issue/compute';
import { OBJECTS_OF_ISSUE_SECTION_LABELS } from '@/lib/objects-of-issue/options';
import { calculateObjectsOfIssueProgress } from '@/lib/objects-of-issue/progress';
import type {
  CompanyLegalReference,
  IpoSetupReference,
  LinkedWorkstreamReferences,
  ObjectsOfIssueSectionId,
  SectionStatus,
} from '@/lib/objects-of-issue/types';
import type { ObjectsOfIssuePayload } from '@/lib/schemas/objects-of-issue';
import { OBJECTS_OF_ISSUE_SECTION_IDS } from '@/lib/schemas/objects-of-issue';

export type ObjectsOfIssueOverviewSummary = {
  isPureOfs: boolean;
  sectionStatuses: Record<ObjectsOfIssueSectionId, SectionStatus>;
  sectionsComplete: number;
  sectionsInProgress: number;
  totalSections: number;
  objectsCount: number;
  netFreshIssueProceeds: string;
  totalEstimatedObjectsCost: string;
  totalAllocatedFromNetProceeds: string;
  gcpPercentageOfFreshIssue: string;
  hasCapexRelevantObjects: boolean;
  hasAcquisitionRelevantObjects: boolean;
  companyReference: CompanyLegalReference;
  assessmentResultLabel: string;
  assessmentSummary: string;
  reconciledChecksCount: number;
  varianceChecksCount: number;
  missingInformationChecksCount: number;
  recommendedNextActions: Array<{ sectionId: ObjectsOfIssueSectionId; label: string }>;
};

export function buildObjectsOfIssueOverviewSummary(
  payload: ObjectsOfIssuePayload,
  ipoReference: IpoSetupReference,
  linkedReferences: LinkedWorkstreamReferences,
): ObjectsOfIssueOverviewSummary {
  const progress = calculateObjectsOfIssueProgress(payload);
  const model = computeObjectsOfIssueModel(payload, ipoReference);
  const assessment = assessObjectsOfIssue(payload, ipoReference, linkedReferences);

  const sectionsInProgress = OBJECTS_OF_ISSUE_SECTION_IDS.filter(
    (id) => progress.sections[id] === 'in_progress',
  ).length;

  const reconciledChecksCount = assessment.criteria.filter(
    (criterion) => criterion.state === 'reconciled',
  ).length;
  const varianceChecksCount = assessment.criteria.filter(
    (criterion) =>
      criterion.state === 'potential_concern' || criterion.state === 'blocked',
  ).length;
  const missingInformationChecksCount = assessment.criteria.filter(
    (criterion) => criterion.state === 'missing_information',
  ).length;

  const recommendedNextActions = OBJECTS_OF_ISSUE_SECTION_IDS.filter(
    (id) => progress.sections[id] !== 'complete',
  )
    .slice(0, 3)
    .map((id) => ({
      sectionId: id,
      label: `Continue ${OBJECTS_OF_ISSUE_SECTION_LABELS[id]}`,
    }));

  return {
    isPureOfs: model.isPureOfs,
    sectionStatuses: progress.sections,
    sectionsComplete: progress.sectionsComplete,
    sectionsInProgress,
    totalSections: progress.totalSections,
    objectsCount: payload.objectsRegisterAndAllocation.objects.length,
    netFreshIssueProceeds: model.netFreshIssueProceeds,
    totalEstimatedObjectsCost: model.totalEstimatedObjectsCost,
    totalAllocatedFromNetProceeds: model.totalAllocatedFromNetProceeds,
    gcpPercentageOfFreshIssue: model.gcpPercentageOfFreshIssue,
    hasCapexRelevantObjects: model.hasCapexRelevantObjects,
    hasAcquisitionRelevantObjects: model.hasAcquisitionRelevantObjects,
    companyReference: linkedReferences.company,
    assessmentResultLabel: assessment.resultLabel,
    assessmentSummary: assessment.summary,
    reconciledChecksCount,
    varianceChecksCount,
    missingInformationChecksCount,
    recommendedNextActions,
  };
}

File: backend/app/modules/objects_issue/overview.py
"""Overview-tab summary for Objects of the Issue — ports overview.ts."""

from __future__ import annotations

from typing import Any

from app.modules.objects_issue.assessment import assess_objects_of_issue
from app.modules.objects_issue.compute import compute_objects_of_issue_model
from app.modules.objects_issue.constants import SECTION_IDS, SECTION_LABELS
from app.modules.objects_issue.progress import calculate_progress

SLUG = "objects-of-issue"


def build_overview_summary(
    payload: dict[str, Any],
    ipo_reference: dict[str, Any],
    linked_references: dict[str, Any],
) -> dict[str, Any]:
    progress = calculate_progress(payload)
    model = compute_objects_of_issue_model(payload, ipo_reference)
    assessment = assess_objects_of_issue(payload, ipo_reference, linked_references)

    sections_in_progress = sum(
        1 for section_id in SECTION_IDS if progress["sections"].get(section_id) == "in_progress"
    )

    reconciled_checks_count = sum(
        1 for criterion in assessment["criteria"] if criterion["state"] == "reconciled"
    )
    variance_checks_count = sum(
        1
        for criterion in assessment["criteria"]
        if criterion["state"] in ("potential_concern", "blocked")
    )
    missing_information_checks_count = sum(
        1 for criterion in assessment["criteria"] if criterion["state"] == "missing_information"
    )

    recommended_next_actions = [
        {
            "sectionId": section_id,
            "label": f"Continue {SECTION_LABELS[section_id]}",
        }
        for section_id in SECTION_IDS
        if progress["sections"].get(section_id) != "complete"
    ][:3]

    company = linked_references.get("company") or {
        "available": False,
        "legalName": None,
        "companyClass": None,
        "cin": None,
    }

    return {
        "isPureOfs": model["isPureOfs"],
        "sectionStatuses": progress["sections"],
        "sectionsComplete": progress["sectionsComplete"],
        "sectionsInProgress": sections_in_progress,
        "totalSections": progress["totalSections"],
        "objectsCount": len((payload.get("objectsRegisterAndAllocation") or {}).get("objects") or []),
        "netFreshIssueProceeds": model["netFreshIssueProceeds"],
        "totalEstimatedObjectsCost": model["totalEstimatedObjectsCost"],
        "totalAllocatedFromNetProceeds": model["totalAllocatedFromNetProceeds"],
        "gcpPercentageOfFreshIssue": model["gcpPercentageOfFreshIssue"],
        "gcpApplicableCap": model["gcpApplicableCap"],
        "hasCapexRelevantObjects": model["hasCapexRelevantObjects"],
        "hasAcquisitionRelevantObjects": model["hasAcquisitionRelevantObjects"],
        "companyReference": company,
        "assessmentResult": assessment["result"],
        "assessmentResultLabel": assessment["resultLabel"],
        "assessmentSummary": assessment["summary"],
        "blockingConcernCount": assessment["metrics"]["blockingConcerns"],
        "reconciledChecksCount": reconciled_checks_count,
        "varianceChecksCount": variance_checks_count,
        "missingInformationChecksCount": missing_information_checks_count,
        "recommendedNextActions": recommended_next_actions,
    }
